File: ok2curl/src/test/java/com/moczul/ok2curl/util/HeaderModifierTest.java
package com.moczul.ok2curl.util;

import com.moczul.ok2curl.CurlBuilder;
import com.moczul.ok2curl.Header;
import com.moczul.ok2curl.Options;
import com.moczul.ok2curl.modifier.HeaderModifier;

import org.junit.Test;

import java.util.Collections;
import java.util.List;

import okhttp3.Request;

import static junit.framework.Assert.assertEquals;

public class HeaderModifierTest {

    private final HeaderModifier cookieHeaderModifier = new HeaderModifier() {

        @Override
        public boolean matches(Header header) {
            return "Cookie".equals(header.name());
        }

        @Override
        public Header modify(Header header) {
            return new Header(header.name(), "modifiedCookieValue");
        }
    };

    private final HeaderModifier nullHeaderModifier = new HeaderModifier() {

        @Override
        public boolean matches(Header header) {
            return true;
        }

        @Override
        public Header modify(Header header) {
            return null;
        }
    };

    @Test
    public void curlCommand_shouldContains_modifiedHeader() {
        final Request request = new Request.Builder()
                .url("http://example.com/")
                .header("Cookie", "FIRST=foo")
                .build();

        final List<HeaderModifier> modifiers = Collections.singletonList(cookieHeaderModifier);
        final String command = new CurlBuilder(request, -1L, modifiers, Options.EMPTY).build();

        assertEquals("curl -X GET -H \"Cookie:modifiedCookieValue\" http://example.com/", command);
    }

    @Test
    public void curlCommand_shouldNotBeModified_ifDoesNotContainMatchingHeader() {
        final Request request = new Request.Builder()
                .url("http://example.com/")
                .header("Accept", "application/json")
                .build();
        final List<HeaderModifier> modifiers = Collections.singletonList(cookieHeaderModifier);

        final String command = new CurlBuilder(request, -1L, modifiers, Options.EMPTY).build();

        assertEquals("curl -X GET -H \"Accept:application/json\" http://example.com/", command);
    }

    @Test
    public void curlCommand_shouldNotContainsAnyHeaders_forNullHeaderModifier() {
        final Request request = new Request.Builder()
                .url("http://example.com/")
                .header("Cookie", "FIRST=foo")
                .header("Accept", "application/json")
                .build();

        final List<HeaderModifier> modifiers = Collections.singletonList(nullHeaderModifier);
        final String command = new CurlBuilder(request, -1L, modifiers, Options.EMPTY).build();

        assertEquals(command, "curl -X GET http://example.com/", command);
    }
}

File: ok2curl/src/test/java/com/moczul/ok2curl/util/HeaderModifierTest.kt
package com.moczul.ok2curl.util

import com.moczul.ok2curl.Configuration
import com.moczul.ok2curl.CurlGenerator
import com.moczul.ok2curl.Header
import com.moczul.ok2curl.modifier.HeaderModifier
import okhttp3.Request
import org.junit.Assert.assertEquals
import org.junit.Test

class HeaderModifierTest {
    private val cookieHeaderModifier: HeaderModifier = object : HeaderModifier {
        override fun matches(header: Header): Boolean {
            return "Cookie" == header.name
        }

        override fun modify(header: Header): Header {
            return Header(header.name, "modifiedCookieValue")
        }
    }
    private val nullHeaderModifier: HeaderModifier = object : HeaderModifier {
        override fun matches(header: Header): Boolean {
            return true
        }

        override fun modify(header: Header): Header? {
            return null
        }
    }

    @Test
    fun `should generate GET command with modified header`() {
        // given
        val config = Configuration(headerModifiers = listOf(cookieHeaderModifier))
        val curlGenerator = CurlGenerator(configuration = config)
        val request: Request = Request.Builder()
            .url("http://example.com/")
            .header("Cookie", "FIRST=foo")
            .build()

        // when
        val command = curlGenerator.generateCommand(request)

        // then
        assertEquals(
            "curl -X GET -H \"Cookie:modifiedCookieValue\" \"http://example.com/\"",
            command
        )
    }

    @Test
    fun `should generate GET command with unmodified header`() {
        // given
        val config = Configuration(headerModifiers = listOf(cookieHeaderModifier))
        val curlGenerator = CurlGenerator(configuration = config)
        val request: Request = Request.Builder()
            .url("http://example.com/")
            .header("Accept", "application/json")
            .build()

        // when
        val command = curlGenerator.generateCommand(request)

        // then
        assertEquals(
            "curl -X GET -H \"Accept:application/json\" \"http://example.com/\"",
            command
        )
    }

    @Test
    fun `should generate GET command without any headers`() {
        // given
        val config = Configuration(headerModifiers = listOf(nullHeaderModifier))
        val curlGenerator = CurlGenerator(configuration = config)
        val request: Request = Request.Builder()
            .url("http://example.com/")
            .header("Cookie", "FIRST=foo")
            .header("Accept", "application/json")
            .build()

        // when
        val command = curlGenerator.generateCommand(request)

        // then
        assertEquals(command, "curl -X GET \"http://example.com/\"", command)
    }
}